### PR TITLE
ScheduleLoad: avoid sinking loads in loop

### LINKED ISSUE #1951: ScheduleLoad transformation sink invariant load in loop
The advanced codegen path uses a transformation (ScheduleLoad.cpp) designed to sink loads of `tt.dot` operands closer to their use. 
While collecting loads to be moved, the transformation collects not just loads in the same region (e.g. loop body) as their uses but also collects loads that are outside of the region (e.g. outside of the loop body). Example:

Before transformation
```
    %33 = tt.load %14 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
    %34 = tt.load %15 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
    %35:24 = scf.for %arg6 = %c0_i32 to %c1024_i32 step %c64_i32 iter_args(%arg7 = %cst, %arg8 = %cst, %arg9 = %cst, %arg10 = %cst, %arg11 = %cst, %arg12 = %cst, %arg13 = %cst, %arg14 = %cst, %arg15 = %17, %arg16 = %18, %arg17 = %19, %arg18 = %20, %arg19 = %21, %arg20 = %22, %arg21 = %23, %arg22 = %24, %arg23 = %25, %arg24 = %26, %arg25 = %27, %arg26 = %28, %arg27 = %29, %arg28 = %30, %arg29 = %31, %arg30 = %32) -> (tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>)  : i32 {
      %44 = tt.load %arg15 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %45 = tt.load %arg16 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %46 = tt.load %arg17 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %47 = tt.load %arg18 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %48 = tt.load %arg19 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %49 = tt.load %arg20 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %50 = tt.load %arg21 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %51 = tt.load %arg22 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %52 = tt.load %arg23 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %53 = tt.load %arg24 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %54 = tt.load %arg25 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %55 = tt.load %arg26 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %56 = tt.load %arg27 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %57 = tt.load %arg28 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %58 = tt.load %arg29 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %59 = tt.load %arg30 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      %60 = triton_intel_gpu.extract %33[0] : tensor<16x32xf16> -> tensor<8x16xf16>
      etc ...
```

After transformation
```
    %32 = tt.make_tensor_ptr %16, [%c64_i64, %c1024_i64], [%c1_i64, %c64_i64], [%c48_i32, %c48_i32] {order = array<i32: 0, 1>} : <tensor<16x16xf16>>
    %33:24 = scf.for %arg6 = %c0_i32 to %c1024_i32 step %c64_i32 iter_args(%arg7 = %cst, %arg8 = %cst, %arg9 = %cst, %arg10 = %cst, %arg11 = %cst, %arg12 = %cst, %arg13 = %cst, %arg14 = %cst, %arg15 = %17, %arg16 = %18, %arg17 = %19, %arg18 = %20, %arg19 = %21, %arg20 = %22, %arg21 = %23, %arg22 = %24, %
arg23 = %25, %arg24 = %26, %arg25 = %27, %arg26 = %28, %arg27 = %29, %arg28 = %30, %arg29 = %31, %arg30 = %32) -> (tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, tensor<8x16xf32>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x
16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<
16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>, !tt.ptr<tensor<16x16xf16>>)  : i32 {
      %42 = tt.load %14 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
      %43 = triton_intel_gpu.extract %42[0] : tensor<16x32xf16> -> tensor<8x16xf16>
      %44 = triton_intel_gpu.extract %42[2] : tensor<16x32xf16> -> tensor<8x16xf16>
      %45 = tt.load %15 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
      %46 = triton_intel_gpu.extract %45[0] : tensor<16x32xf16> -> tensor<8x16xf16>
      %47 = triton_intel_gpu.extract %45[2] : tensor<16x32xf16> -> tensor<8x16xf16>
      %48 = triton_intel_gpu.extract %42[1] : tensor<16x32xf16> -> tensor<8x16xf16>
      %49 = triton_intel_gpu.extract %42[3] : tensor<16x32xf16> -> tensor<8x16xf16>
      %50 = triton_intel_gpu.extract %45[1] : tensor<16x32xf16> -> tensor<8x16xf16>
      %51 = triton_intel_gpu.extract %45[3] : tensor<16x32xf16> -> tensor<8x16xf16>
      %52 = tt.load %arg15 {DotIdx = 1 : i32} : !tt.ptr<tensor<16x16xf16>>
      etc...
```

Note that the 2 loads that were defined outside the loop:
```
 %33 = tt.load %14 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
 %34 = tt.load %15 {DotIdx = 0 : i32} : !tt.ptr<tensor<16x32xf16>>
 ```
Have been "sinked" into the loop.

This has the effect of reducing the live range of these 2 loads, but at the cost of performing the memory access at every loop iteration. In the absence of a cost model I believe this transformation should limit itself to scheduling loads in the region where they originally appeared.

### PR DESCRIPTION
**Pull request by etiotto**
Fixes #1951

### CODE REVIEW

File: test/TritonIntelGPU/schedule-load.mlir [hunk 1]
```diff
@@ -1,6 +1,7 @@
-// RUN: TRITON_INTEL_ENABLE_INSTR_SCHED=1 triton-opt %s -split-input-file -tritonintelgpu-schedule-load | FileCheck %s
-module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
-  tt.func public @_attn_fwd(%arg0: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg3: f32, %arg4: !tt.ptr<f32> {tt.divisibility = 16 : i32}, %arg5: !tt.ptr<f32> {tt.divisibility = 16 : i32}) attributes {noinline = false} {
+// RUN: TRITON_INTEL_ENABLE_INSTR_SCHED=1 triton-opt %s -tritonintelgpu-schedule-load | FileCheck %s
+module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
```
Comment (etiotto) [inline test/TritonIntelGPU/schedule-load.mlir:2]: Note: removed several unnecessary attributes and operations to simplify the lit test a bit.

File: third_party/intel/lib/Analysis/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 add_triton_library(TritonIntelAnalysis
     DPAS.cpp
+    Liveness.cpp
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/CMakeLists.txt:3]: Remove, this doesn't belong in this PR.

File: third_party/intel/lib/TritonIntelGPUTransforms/ScheduleLoad.cpp [hunk 3]
```diff
@@ -1,4 +1,4 @@
-//===- ScheduleLoad.cpp ----------------------------------------------*-===//
+//===- ScheduleLoad.cpp -------------------------------------------------*-===//
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/ScheduleLoad.cpp:1]: remove this doesn't belong in this PR

File: third_party/intel/lib/Analysis/CMakeLists.txt (hunk 2 above)
Comment (etiotto) [reply third_party/intel/lib/Analysis/CMakeLists.txt:3]: removed

File: third_party/intel/lib/TritonIntelGPUTransforms/ScheduleLoad.cpp [hunk 4]
```diff
@@ -75,22 +84,23 @@ class ScheduleLoadPass
...
-          auto op = val.getDefiningOp();
-          if (i == 0)
-            op->moveBefore(start);
-          else
-            op->moveBefore(dots.begin()->getOperation());
+          Operation *insertPoint =
+              (groupNum == 0) ? start
+                              : dots.begin()->getOperation()->getPrevNode();
+
+          if (canMoveValueTo(val, insertPoint))
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/ScheduleLoad.cpp:98]: Added check to ensure the move is legal.

**Comment by LiyangLingIntel** [comment]
For FlashAttention case, I think we do need sinking loads in the loop.  
@Dewei-Wang-sh Can you please take a look?

**Comment by etiotto** [comment]
> For FlashAttention case, I think we do need sinking loads in the loop. @Dewei-Wang-sh Can you please take a look?

If that is the case then I can create a env. variable to allow the pass to sink loads in the loop.  This is temporary because we want to build a cost model (based on live intervals analysis) to guide this decision.

**Comment by etiotto** [comment]
Decided to split this PR and place the changes to reduce the lit test separately.

**Review by victor-eds — APPROVED** [review]
